nixos/modules/services/matrix/conduwuit.nix

```diff
@@ -0,0 +1,165 @@
+            type = lib.types.str;
+            default = "/var/lib/conduwuit/";
+            readOnly = true;
+            description = ''
+              Path to the conduwuit database, the directory where conduwuit will save its data.
+              Note that due to using the DynamicUser feature of systemd, this value should not be changed
+              and is set to be read only.
+            '';
+          };
+          global.database_backend = lib.mkOption {
+            type = lib.types.enum [ "rocksdb" ];
+            default = "rocksdb";
+            example = "rocksdb";
+            description = ''
+              The database backend for the service. Switching it on an existing
+              instance will require manual migration of data.
+
+              Only rocksdb is supported.
+            '';
+          };
+          global.allow_check_for_updates = lib.mkOption {
+            type = lib.types.bool;
+            default = false;
+            description = ''
+              If enabled, conduwuit will send a simple GET request periodically to
+              <https://pupbrain.dev/check-for-updates/stable> for any new announcements made.
+              Despite the name, this is not an update check endpoint, it is simply an announcement check endpoint.
+
+              Disabled by default.
+            '';
+          };
+        };
+      };
+      default = { };
+      description = ''
+        Generates the conduwuit.toml configuration file. Refer to
+        <https://conduwuit.puppyirl.gay/configuration.html>
+        for details on supported values.
+        Note that database_path can not be edited because the service's reliance on systemd StateDir.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.conduwuit = {
+      description = "Conduwuit Matrix Server";
+      documentation = [ "https://conduwuit.puppyirl.gay/" ];
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network-online.target" ];
+      after = [ "network-online.target" ];
+      environment = lib.mkMerge ([
+        { CONDUWUIT_CONFIG = configFile; }
+        cfg.extraEnvironment
+      ]);
+      serviceConfig = {
+        DynamicUser = true;
+        User = "conduwuit";
+        LockPersonality = true;
+        MemoryDenyWriteExecute = true;
+        ProtectClock = true;
+        ProtectControlGroups = true;
+        ProtectHostname = true;
+        ProtectKernelLogs = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        PrivateDevices = true;
+        PrivateMounts = true;
+        PrivateUsers = true;
+        RestrictAddressFamilies = [
+          "AF_INET"
+          "AF_INET6"
+        ];
+        RestrictNamespaces = true;
+        RestrictRealtime = true;
+        SystemCallArchitectures = "native";
+        SystemCallFilter = [
+          "@system-service"
+          "~@privileged"
+        ];
+        StateDirectory = "conduwuit";
```

Comment:
Could we also add `RuntimeDirectory=conduwuit` and `RuntimeDirectoryMode=0750` to allow users to place the UNIX socket in `/run/conduwuit/conduwuit.sock` and give any systemd units or users permission to read/write the socket with the `conduwuit` group? This is how I use UNIX sockets with conduwuit for slightly increased host security.

Obviously nothing is stopping a user from putting the socket in `/tmp` or something.

Reply:
I would prefer not to set `RuntimeDirectoryMode=0750` by default since this would require users to either override this option or to statically allocate the `conduwuit` user/group if they wish to put the unix socket in `RuntimeDirectory`.

Users can easily set `RuntimeDirectoryMode=` independent of this module if they wish to further restrict access to the unix sockt.

Reply:
> or to statically allocate the conduwuit user/group if they wish to put the unix socket in RuntimeDirectory.

This is how we recommend the usage of UNIX sockets in conduwuit, and technically the proper way to use them. The default `unix_socket_perms` config option in conduwuit is `660` and our example config documentation tells you to add the necessary `conduwuit` group to your reverse proxy.

https://conduwuit.puppyirl.gay/configuration/examples.html#example-configuration

```
# Uncomment unix_socket_path to listen on a UNIX socket at the specified
# path. If listening on a UNIX socket, you MUST remove/comment the
# 'address' key if definedm AND add your reverse proxy to the 'conduwuit'
# group, unless world RW permissions are specified with unix_socket_perms
# (666 minimum).
#
# example: "/run/conduwuit/conduwuit.sock"
#
#unix_socket_path =

# The default permissions (in octal) to create the UNIX socket with.
#
#unix_socket_perms = 660
```

Our example systemd unit also has this set: https://conduwuit.puppyirl.gay/configuration/examples.html#debian-systemd-unit-file

If users choose to use UNIX sockets differently, they're welcome to, but the more secure and proper way to use them is confining them.

I don't think the module should deviate from conduwuit's default and recommendations here.

Reply:
Okay then maybe the module should reflect that and automatically allocate a user/group and disable DynamicUser if `unix_socket_path != null`

Reply:
It is actually quite easy I'll just write a suggestion